generators/protoc-gen-go-googlecetypes/generator.go

```diff
@@ -0,0 +1,156 @@
+	// It is mapped to the DataType.
+	PrepareFunction map[string]string
+
+	// Protoc tooling versions.
+	ProtocVersion      string
+	ProtocGenGoVersion string
+	LibraryVersion     string
+}
+
+// generateTests generates test coverage per type.
+func generateTests(gen *protogen.Plugin, file *protogen.File) *protogen.GeneratedFile {
+	filename := filepath.Join(filepath.Dir(file.GeneratedFilenamePrefix), "data_test.go")
+	g := gen.NewGeneratedFile(filename, file.GoImportPath)
+
+	params := TestParams{
+		Pkg:        string(file.GoPackageName),
+		SrcPath:    strings.TrimPrefix(strings.Trim(file.GoImportPath.String(), "\""), SrcPrefix),
+		TypePrefix: string(file.Desc.Package()),
+		SourceFile: file.Desc.Path(),
+	}
+
+	// Derive the protoc version.
+	protocVersion := "(unknown)"
+	if v := gen.Request.GetCompilerVersion(); v != nil {
+		protocVersion = fmt.Sprintf("v%v.%v.%v", v.GetMajor(), v.GetMinor(), v.GetPatch())
+	}
+	params.ProtocVersion = protocVersion
+
+	// Derive protoc-gen-go version.
+	pgcVersion := "(unknown)"
+	if v := os.Getenv("PROTOC_GEN_GO_VERSION"); v != "" {
+		pgcVersion = v
+	}
+	params.ProtocGenGoVersion = pgcVersion
+
+	// Derive current library version.
+	libVersion := "(unknown)"
+	if v := os.Getenv("LIBRARY_VERSION"); v != "" {
+		libVersion = v
+	}
+	params.LibraryVersion = libVersion
+
+	// Notes on parsing a proto message:
+	// msg.GoIdent.GoName => Event Name like "FunctionCreatedEvent"
+	// msg.Desc.FullName() => CloudEvent Type like "google.events.cloud.functions.v2.FunctionCreatedEvent"
+	// msg.Desc.Fields().ByName("data").Message().FullName() => Event Data Type like "FunctionEventData"
+
+	// Assemble DataTypes
+	dataTypeMap := make(map[string]string, len(file.Messages))
+	for _, msg := range file.Messages {
+		dataType := filepath.Ext(string(msg.Desc.Fields().ByName("data").Message().FullName()))[1:]
+		dataTypeMap[dataType] = msg.GoIdent.GoName
+	}
+	var dataTypeSlice []string
+	for i := range dataTypeMap {
+		dataTypeSlice = append(dataTypeSlice, i)
+	}
+	params.DataTypes = dataTypeSlice
+
+	// Create TestDataPath like "google/events/cloud/functions/v1"
+	re := regexp.MustCompile(`(v\d)$`)
+	version := re.FindString(string(file.GoPackageName))
+	product := strings.TrimSuffix(string(file.GoPackageName), version)
+	product = strings.TrimSuffix(product, "data")
+	if version == "" {
+		version = "v1"
+	}
+
+	testDataPath := filepath.Join(SrcPrefix, filepath.Dir(params.SrcPath), product, version)
+	params.TestDataPath = testDataPath
+
+	// Identify if a data cleaning step is required and assign the cleanup function.
+	params.PrepareFunction = make(map[string]string, len(params.DataTypes))
+	for _, dataType := range params.DataTypes {
+		switch {
+		case params.TypePrefix == "google.events.cloud.pubsub.v1" && dataType == "MessagePublishedData":
+			params.PrepareFunction[dataType] = "testhelper.PreparePubSubMessagePublishedData"
```

Comment:
What was your reasoning for choosing strings here rather than passing the function?

IMO, the latter will help avoid typos and should be simpler to call from the template, but there may be other reasons not to do that.

Reply:
Not sure I'm following. The function being called is in the internal package, it's not defined or run at generation time.

Reply:
The string in this map (`"testhelper.PreparePubSubMessagePublishedData"`) is referring to a function. This function is used during the generation process when filling out the template, right? If there's a typo in it, this will only be apparent when the generator runs.

if you use a reference to the function itself (ie. import the testhelper and use just `testhelper.PreparePubSubMessagePublishedData` without quotes), typos in the string will be found at build time (as undefined symbol), rather than run time. IMO, this is a more maintainable approach.